File: sparke-Java/src/main/java/com/SparkEngine/InitSpark.java
package com.SparkEngine;

/**
 * Created by alikemal on 19.03.2017.
 */

import com.Conf;
import com.model.Song;
import com.model.Tag;
import com.model.TasteStat;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.apache.spark.api.java.StorageLevels.MEMORY_ONLY;

public class InitSpark {
    public static SparkSession spark = null;
    private final Logger LOGGER = LoggerFactory.getLogger(InitSpark.class);

    private final Dataset<Row> userJson;
    private final Dataset<Row> songJson;
    private final Dataset<Row> tagJson;
/*
    private final Dataset<Row> tagParq;
    private final Dataset<Row> songParq;
    private final Dataset<Row> tasteParq;*/
    private final Dataset<Row> ratingJson;

    public InitSpark() {

        SparkConf conf = new SparkConf()
                .setAppName(InitSpark.class.getName())
                .setMaster("local");
        spark = SparkSession
                .builder()
                .config(conf)
                .getOrCreate();


        //******************************************

        userJson = spark.read().json(Conf.usersJson_txt);
        songJson = spark.read().json(Conf.songJson_txt);
        tagJson = spark.read().json(Conf.tagJson_txt);
        ratingJson = spark.read().json(Conf.ratingJson_txt);

        /*tagParq = spark.read().parquet(Conf.tag_parqPATH);
        tasteParq = spark.read().parquet(Conf.taste_parqPATH);
        songParq = spark.read().parquet(Conf.song_parqPATH);


        /*//******************************************
        songParq.createOrReplaceTempView("song");

        tasteParq.createOrReplaceTempView("tastestats");

        tagParq.createOrReplaceTempView("tag");*/

        userJson.createOrReplaceTempView("userjson");

        songJson.createOrReplaceTempView("songjson");

        tagJson.createOrReplaceTempView("tagjson");

        ratingJson.createOrReplaceTempView("ratingjson");

        //*****************************************


    }

    public static String generateParquet() {

        final JavaRDD<String> songRead = spark.read().textFile(Conf.tracks_txt).toJavaRDD();
        final JavaRDD<String> tasteStatRead = spark.read().textFile(Conf.tasteStat_Txt).toJavaRDD();
        final JavaRDD<String> tagRead = spark.read().textFile(Conf.tag_txt).javaRDD();

        JavaRDD<Song> songRDD = songRead
                .map(lineRAW -> {
                    String[] parts = lineRAW.split(",");
                    return new Song(parts[0], parts[1], parts[2], parts[3]);
                }).persist(MEMORY_ONLY).cache();

        Dataset<Row> songDF = spark.createDataFrame(songRDD, Song.class);
        songDF.write().parquet(Conf.song_parqPATH);

        //******************************************

        JavaRDD<TasteStat> tasteRDD = tasteStatRead.map(lineRAW -> {
            String[] parts = lineRAW.split(",");
            return new TasteStat(parts[1], parts[0], Integer.parseInt(parts[2]));
        });

        Dataset<Row> tasteDF = spark.createDataFrame(tasteRDD, TasteStat.class);
        tasteDF.write().parquet(Conf.taste_parqPATH);

        //******************************************

        JavaRDD<Tag> tagRDD = tagRead
                .map(lineRAW -> {
                    String[] parts = lineRAW.split(",");
                    return new Tag(parts[0], parts[1]);
                }).persist(MEMORY_ONLY).cache();

        Dataset<Row> tagDF = spark.createDataFrame(tagRDD, Tag.class);
        tagDF.write().parquet(Conf.tag_parqPATH);
        return "ok";
    }


}

File: InitSpark.py
from pyspark import SparkConf, SparkContext, StorageLevel, Row
from pyspark.sql import SparkSession

conf = (SparkConf()
        .set("spark.executor.cores", "12")
        .set("spark.executor.memory", "5g")
        .setAppName("Spark-API")
        .setMaster("local[*]"))


# master="spark://alikemal-300E5C:7077"
#  spark://192.168.2.106:7077
# custom_schema = StructType({
#     StructField("userid", IntegerType(), False),
#     StructField("songid", IntegerType(), False),
#     StructField("rating", LongType(), False)})


class InitSpark:
    def __init__(self):
        self.sc = SparkContext(conf=conf)
        locale = self.sc._jvm.java.util.Locale
        locale.setDefault(locale.forLanguageTag("tr-TR"))
        self.sc.setLogLevel("INFO")
        self.spark = SparkSession(self.sc) \
            .builder \
            .config("spark.mongodb.input.uri",
                    "mongodb://alikemal:123456alikemal@alikemal.org:27017/admin.recommend?authMechanism=SCRAM-SHA-1") \
            .config("spark.mongodb.output.uri",
                    "mongodb://alikemal:123456alikemal@alikemal.org:27017/admin.recommend?authMechanism=SCRAM-SHA-1") \
            .config("packages", "org.mongodb.spark:mongo-spark-connector_2.11:2.0.0") \
            .getOrCreate()

        self.base_txt = "dataset/"
        self.usercsv_txt = self.base_txt + "userid-profile.tsv"

        self.ratingParq = self.base_txt + "rating.json"
        self.song_parqPATH = self.base_txt + "songs.parquet"

        # userJson = self.spark.read.json(self.usersJson_txt).persist(StorageLevel.MEMORY_ONLY).cache() \
        #     .createOrReplaceTempView("userjson")

        self.spark.read.load(self.ratingParq, format='json').persist(StorageLevel.MEMORY_ONLY).cache() \
            .createOrReplaceTempView("rating")
        self.spark.read.parquet(self.song_parqPATH).persist(StorageLevel.MEMORY_ONLY) \
            .createOrReplaceTempView("song")

    def generateParquet(self):
        # track_line = self.sc.textFile(self.train_txt)
        # songs = track_line.map(lambda l: l.split(",")) \
        #     .map(lambda p: Row(trackID=p[0], songID=p[1], artistName=p[2], songTitle=p[3]))
        #
        # tag_line = self.sc.textFile(self.tag_txt)
        # tags = tag_line.map(lambda l: l.split(",")) \
        #     .map(lambda p: Row(trackID=p[0], tagID=p[1]))
        #
        # schemaSongs = self.spark.createDataFrame(songs)
        # schemaSongs.write.parquet(self.song_parqPATH)


        # track_line = self.sc.textFile(self.train_txt)
        # songs = track_line.map(lambda l: l.split(",")) \
        #     .map(lambda p: Row(tuserID=p[0], songID=p[1], rating=float(p[2]))).take(1000000)
        #
        # schemaSongs = self.spark.createDataFrame(songs)
        # schemaSongs.createOrReplaceTempView("train")
        #
        # self.spark.sql("SELECT "
        #                "DENSE_RANK() OVER ( ORDER BY t.tuserID) as userid,s.sarkiId as songid,t.rating,s.genreID as genreID "
        #                       " FROM song as s "
        #                       "   INNER JOIN train as t ON "
        #                       "s.songID=t.songID").write.save(self.base_txt + "rating.json", format='json')
        # return "ok"


        track_line = self.sc.textFile(self.base_txt + "userid-timestamp-artid-artname-traid-traname.tsv")
        songs = track_line.map(lambda l: l.split("\t")) \
            .map(lambda p: Row(user=str(p[0]), artname=str(p[3]), traname=str(p[5]))) \
            .take(1000000)

        # Infer the schema, and register the DataFrame as a table.
        schemaSongs = self.spark.createDataFrame(songs)
        schemaSongs.show()
        schemaSongs.write.save(self.base_txt + "1ksong.json", format='json')
